pytorch_lightning/trainer/lr_finder.py

```diff
@@ -0,0 +1,434 @@
+            self.progress_bar = tqdm(desc='Finding best initial lr', total=self.num_training)
+
+        self.lrs.append(trainer.lr_schedulers[0]['scheduler'].lr[0])
+
+    def on_batch_end(self, trainer, pl_module):
+        """ Called when the training batch ends, logs the calculated loss """
+        if self.progress_bar:
+            self.progress_bar.update()
+
+        current_loss = trainer.running_loss.last().item()
+        current_step = trainer.global_step + 1  # remove the +1 in 1.0
+
+        # Avg loss (loss with momentum) + smoothing
+        self.avg_loss = self.beta * self.avg_loss + (1 - self.beta) * current_loss
+        smoothed_loss = self.avg_loss / (1 - self.beta**current_step)
+
+        # Check if we diverging
+        if current_step > 1 and smoothed_loss > 4 * self.best_loss:
+            trainer.max_steps = current_step  # stop signal
+            if self.progress_bar:
+                self.progress_bar.close()
+
+        # Save best loss for diverging checking
+        if smoothed_loss < self.best_loss or current_step == 1:
+            self.best_loss = smoothed_loss
+
+        self.losses.append(smoothed_loss)
+
+
+class _LinearLR(_LRScheduler):
+    """Linearly increases the learning rate between two boundaries
+    over a number of iterations.
+    Arguments:
+
+        optimizer: wrapped optimizer.
+
+        end_lr: the final learning rate.
+
+        num_iter: the number of iterations over which the test occurs.
+
+        last_epoch: the index of last epoch. Default: -1.
+    """
+
+    def __init__(self,
+                 optimizer: torch.optim.Optimizer,
+                 end_lr: float,
+                 num_iter: int,
+                 last_epoch: int = -1):
+        self.end_lr = end_lr
+        self.num_iter = num_iter
+        super(_LinearLR, self).__init__(optimizer, last_epoch)
+
+    def get_lr(self):
+        curr_iter = self.last_epoch + 1
+        r = curr_iter / self.num_iter
+
+        if self.last_epoch > 0:
+            val = [base_lr + r * (self.end_lr - base_lr) for base_lr in self.base_lrs]
+        else:
+            val = [base_lr for base_lr in self.base_lrs]
+        self._lr = val
+        return val
+
+    @property
+    def lr(self):
+        return self._lr
+
+
+class _ExponentialLR(_LRScheduler):
+    """Exponentially increases the learning rate between two boundaries
+    over a number of iterations.
+
+    Arguments:
+
+        optimizer: wrapped optimizer.
+
+        end_lr: the final learning rate.
+
+        num_iter: the number of iterations over which the test occurs.
+
+        last_epoch: the index of last epoch. Default: -1.
+    """
+
+    def __init__(self,
+                 optimizer: torch.optim.Optimizer,
+                 end_lr: float,
+                 num_iter: int,
+                 last_epoch: int = -1):
+        self.end_lr = end_lr
+        self.num_iter = num_iter
+        super(_ExponentialLR, self).__init__(optimizer, last_epoch)
+
+    def get_lr(self):
```

Comment:
it shall be described what is the doff between this `get_lr` and just `lr` bellow because intuitively (by the name) they shall return the same

Reply:
`get_lr()` is the method called inside `lr_scheduler.step()` and is not meant to be called elsewhere. Since pytorch 1.4 the property `self._last_lr` was introduced to extract the last computed lr. However, since pytorch-ligning need to be backwards compatible, I created the `self.lr` property that archives the same. They therfore have slightly different purpose.